### PR TITLE
SIDM-7178 - Removing SSO

### PR DESCRIPTION
**Pull request by justiceia**
### JIRA link ###
https://tools.hmcts.net/jira/browse/SIDM-7178

### Change description ###
DO NOT MERGE IN UNTIL API IS UPDATED TO ALLOW REMOVAL OF SSO

- Added option to remove single sign on option from the users information page
- once remove the single sign on fields should simply show as blank
- updated to use feature flags

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/controllers/UserResultsController.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ export class UserResultsController extends RootController {
         return super.post(req, res, 'user-details', {
           content: {
             user,
-            showDelete: this.canDeleteUser(req.session.user, user),
+            showManage: this.canManageUser(req.session.user, user),
```
Comment (KianKwa) [inline src/main/controllers/UserResultsController.ts:42]: 'showManage' sounds a bit weird. May be show rename it to 'canManage'?

File: src/main/controllers/UserSsoController.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+@autobind
+export class UserSsoController extends RootController{
+
+  @asyncError
+  public post(req: AuthedRequest, res: Response) {
+    return req.scope.cradle.api.getUserById(req.body._userId)
+      .then(user => {
+        switch(req.body.confirmSso) {
+          case 'true':
+            return this.removesso(req, res, user);
```
Comment (KianKwa) [inline src/main/controllers/UserSsoController.ts:22]: Non standard casing for method name. It should be removeSso() and not removesso()

File: src/main/views/sso-user-successful.njk [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  {{ breadcrumb([
+    ['Manage users', urls.MANAGER_USERS_URL],
+    ['Remove SSO']
+  ]) }}
+{% endblock %}
+
+{% block content %}
+  <div class="govuk-grid-row">
+      <div class="govuk-grid-column-three-quarters">
+        {{ govukPanel({ titleText: "Single sign on has been removed" }) }}
```
Comment (KianKwa) [inline src/main/views/sso-user-successful.njk:21]: Think it should be 'Single sign-on' not 'Single sign on'

File: src/main/views/sso-user-successful.njk [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+        <div class="govuk-!-margin-bottom-9">
+          <p class="govuk-body-m govuk-!-margin-bottom-0">Single sign on has been removed for the following account: </p>
```
Comment (KianKwa) [inline src/main/views/sso-user-successful.njk:24]: Should use 'Single sign-on'

File: src/main/views/sso-user-successful.njk [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{% from "macros/breadcrumb.njk" import breadcrumb %}
```
Comment (KianKwa) [inline src/main/views/sso-user-successful.njk:1]: The name of this file should be 'remove-user-sso-successful.njk' stating the action performed. 'sso-user-successful' sounds like they successfully use SSO

File: src/main/views/sso-user-successful.njk (hunk 3 above)
Comment (KianKwa) [reply src/main/views/sso-user-successful.njk:21]: Also the title should be 'Single sign-on removed successfully' to keep it consistent with 'delete user successful' and 'remove user successful' views

File: src/main/views/sso-user.njk [hunk 6]
```diff
@@ -0,0 +1,60 @@
+{% from "macros/breadcrumb.njk" import breadcrumb %}
```
Comment (KianKwa) [inline src/main/views/sso-user.njk:1]: Should rename this file to 'remove-user-sso.njk'

File: src/main/views/sso-user.njk [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+          name: "confirmSso",
+          fieldset: {
+            legend: {
+              text: "Are you sure you want to remove single sign on for " + content.user.forename + " " + content.user.surname + "?",
+              isPageHeading: true,
+              classes: "govuk-fieldset__legend--l"
+            }
+          },
+          hint: {
+            text: "Are you sure you want to remove single sign on for this user? The user will need to reset their password."
```
Comment (KianKwa) [inline src/main/views/sso-user.njk:38]: Should be 'single sign-on' with an hyphen

File: src/main/views/sso-user.njk [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      {{ csrfProtection(csrfToken) }}
+      <input type="hidden" name="_userId" value="{{ content.user.id }}">
+      <div class="govuk-grid-column-one-half">
+        {{ govukRadios({
+          classes: "govuk-radios--inline",
+          idPrefix: "confirmRadio",
+          name: "confirmSso",
+          fieldset: {
+            legend: {
+              text: "Are you sure you want to remove single sign on for " + content.user.forename + " " + content.user.surname + "?",
```
Comment (KianKwa) [inline src/main/views/sso-user.njk:32]: Should be 'single sign-on' with an hyphe

File: src/main/app/feature-flags/flags.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@ export const BETA_ADD = 'idam-user-dashboard--beta-add';
 export const BETA_EDIT = 'idam-user-dashboard--beta-edit';
 export const BETA_SUSPEND = 'idam-user-dashboard--beta-suspend';
 export const BETA_DELETE = 'idam-user-dashboard--beta-delete';
+export const BETA_SSO = 'idam-user-dashboard--beta-sso';
```
Comment (KianKwa) [inline src/main/app/feature-flags/flags.ts:5]: This should be BETA_REMOVE_SSO. This name of the string  should be 'idam-user-dashboard--beta-remove-sso'

File: src/main/app/idam-api/IdamAPI.ts [hunk 10]
```diff
@@ -74,6 +74,17 @@ export class IdamAPI {
       });
   }
 
+  public deleteSsoById(id: string) {
+    return this.userAxios
+      .delete('/api/v1/users/' + id + '/sso')
+      .catch(error => {
+        const errorMessage = 'Error removing user sso by ID from IDAM API';
```
Comment (KianKwa) [inline src/main/app/idam-api/IdamAPI.ts:81]: The error says remove sso by the method name is delete sso. They need to be made consistent

File: src/main/controllers/UserSsoController.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+      });
+  }
+
+  private removesso(req: AuthedRequest, res: Response, user: User) {
+    return req.scope.cradle.api.deleteSsoById(req.body._userId)
+      .then(() => {
+        return super.post(req, res, 'sso-user-successful', { content: { user } } );
+      })
+      .catch(() => {
+        const error = { userRemoveSsoForm: { message: USER_DISABLE_SSO_ERROR } };
```
Comment (KianKwa) [inline src/main/controllers/UserSsoController.ts:44]: Rename the error message to USER_REMOVE_SSO_ERROR to match userRemoveSsoForm

File: src/main/views/user-details.njk [hunk 12]
```diff
@@ -165,8 +165,12 @@
           {{ govukButton({ text: "Unsuspend user", name: "_action", value: "unsuspend", classes: "govuk-button--secondary" }) if not content.user.active }}
         {% endif %}
 
+        {% if featureFlags.values[featureFlags.flags.BETA_SSO] and (content.user.ssoProvider or content.user.ssoId) and content.showManage %}
+          {{ govukButton({ text: "Disable SSO", name: "_action", value: "disable-sso", classes: "govuk-button--secondary" })}}
```
Comment (KianKwa) [inline src/main/views/user-details.njk:169]: Think the ticket says it should be call 'Remove SSO' and not 'Disable SSO'. This endpoint is to remove the SSO info from the user

File: src/test/unit/test/controllers/UserSsoController.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    when(mockApi.getUserById).calledWith(userData.id).mockReturnValue(Promise.resolve(userData));
+    req.body = { _userId: userData.id, _action: 'confirm-remove-sso' };
+
+    await controller.post(req, res);
+    expect(mockApi.getUserById).toBeCalledWith(userData.id);
+    expect(res.render).toBeCalledWith('sso-user', { content: { user: userData }, error });
+  });
+
+  test('Should render the remove sso page after there was an API issue', async () => {
```
Comment (KianKwa) [inline src/test/unit/test/controllers/UserSsoController.ts:93]: The description of test should be 'Should render the remove sso page **with error**......'

Comment (justiceia) [reply src/test/unit/test/controllers/UserSsoController.ts]: This description is consistent with a similar test on other functionality like delete

File: src/main/views/user-details.njk (hunk 12 above)
Comment (justiceia) [reply src/main/views/user-details.njk]: fixed

File: src/main/controllers/UserSsoController.ts (hunk 11 above)
Comment (justiceia) [reply src/main/controllers/UserSsoController.ts]: fixed

File: src/main/controllers/UserSsoController.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import { AuthedRequest } from '../interfaces/AuthedRequest';
```
[the remark this answers is not in the document]
Comment (justiceia) [reply src/main/controllers/UserSsoController.ts]: fixed

File: src/main/app/idam-api/IdamAPI.ts (hunk 10 above)
Comment (justiceia) [reply src/main/app/idam-api/IdamAPI.ts]: fixed

File: src/main/app/feature-flags/flags.ts (hunk 9 above)
Comment (justiceia) [reply src/main/app/feature-flags/flags.ts]: fixed

File: src/main/views/sso-user.njk (hunk 7 above)
Comment (justiceia) [reply src/main/views/sso-user.njk]: fixed

File: src/main/views/sso-user.njk (hunk 6 above)
Comment (justiceia) [reply src/main/views/sso-user.njk]: fixed

File: src/main/views/sso-user-successful.njk (hunk 5 above)
Comment (justiceia) [reply src/main/views/sso-user-successful.njk]: fixed

File: src/main/views/sso-user-successful.njk (hunk 3 above)
Comment (justiceia) [reply src/main/views/sso-user-successful.njk]: fixed

File: src/main/controllers/UserSsoController.ts (hunk 2 above)
Comment (justiceia) [reply src/main/controllers/UserSsoController.ts]: fixed

File: src/main/controllers/UserResultsController.ts (hunk 1 above)
Comment (justiceia) [reply src/main/controllers/UserResultsController.ts]: fixed

File: src/main/app/idam-api/IdamAPI.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (justiceia) [reply src/main/app/idam-api/IdamAPI.ts]: fixed

**Review by KianKwa — APPROVED** [review]

**Pull request merged**